pkg/cache/userpermission.go

```diff
@@ -0,0 +1,1107 @@
+	}
+	if binding.Subject != nil {
+		return []rbacv1.Subject{*binding.Subject}, true
+	}
+	return nil, false
+}
+
+// extractSubjectsFromRoleBinding extracts subjects from a RoleBinding
+// Handles both Subject (singular, embedded) and Subjects (plural, array) fields
+// Per API docs: "If both subject and subjects exist then only subjects will be used"
+func extractSubjectsFromRoleBinding(binding *clusterpermissionv1alpha1.RoleBinding) ([]rbacv1.Subject, bool) {
+	if len(binding.Subjects) > 0 {
+		return binding.Subjects, true
+	}
+	if binding.Subject != nil {
+		return []rbacv1.Subject{*binding.Subject}, true
+	}
+	return nil, false
+}
+
+// processClusterRoleBinding processes a ClusterRoleBinding and adds permissions to the maps
+func (c *UserPermissionCache) processClusterRoleBinding(
+	binding *clusterpermissionv1alpha1.ClusterRoleBinding,
+	clusterName string,
+	userPermissions map[string]map[string][]clusterviewv1alpha1.ClusterBinding,
+	groupPermissions map[string]map[string][]clusterviewv1alpha1.ClusterBinding,
+) {
+	if binding == nil || binding.RoleRef.Name == "" {
+		klog.V(4).Info("Skipping nil or empty ClusterRoleBinding")
+		return
+	}
+
+	roleRefName := binding.RoleRef.Name
+	klog.V(4).Infof("Processing ClusterRoleBinding for role %s in cluster %s", roleRefName, clusterName)
+
+	if !c.discoverableRolesNames.Has(roleRefName) {
+		klog.V(4).Infof("Skipping ClusterRoleBinding: role %s is not in discoverable roles set", roleRefName)
+		return
+	}
+
+	clusterBinding := clusterviewv1alpha1.ClusterBinding{
+		Cluster:    clusterName,
+		Scope:      clusterviewv1alpha1.BindingScopeCluster,
+		Namespaces: []string{"*"},
+	}
+
+	subjects, ok := extractSubjectsFromClusterRoleBinding(binding)
+	if !ok {
+		klog.V(4).Infof("Skipping ClusterRoleBinding for role %s: no subjects found", roleRefName)
+		return
+	}
+
+	klog.V(4).Infof("Adding ClusterRoleBinding for discoverable role %s with %d subject(s)",
+		roleRefName, len(subjects))
+
+	for _, subject := range subjects {
+		switch subject.Kind {
+		case rbacv1.UserKind:
+			addPermissionForUser(userPermissions, subject.Name, roleRefName, clusterBinding)
+		case rbacv1.GroupKind:
+			addPermissionForGroup(groupPermissions, subject.Name, roleRefName, clusterBinding)
+		}
+	}
+}
+
+// processRoleBinding processes a RoleBinding and adds permissions to the maps
+func (c *UserPermissionCache) processRoleBinding(
+	binding *clusterpermissionv1alpha1.RoleBinding,
+	clusterName string,
+	userPermissions map[string]map[string][]clusterviewv1alpha1.ClusterBinding,
+	groupPermissions map[string]map[string][]clusterviewv1alpha1.ClusterBinding,
+) {
+	if binding == nil || binding.RoleRef.Name == "" {
+		klog.V(4).Info("Skipping nil or empty RoleBinding")
+		return
+	}
+
+	roleRefName := binding.RoleRef.Name
+	klog.V(4).Infof("Processing RoleBinding for role %s in cluster %s, namespace %s",
+		roleRefName, clusterName, binding.Namespace)
+
+	if !c.discoverableRolesNames.Has(roleRefName) {
```

Comment:
why we need this check? The discoverableRole is only for clusterPermission I think.